### ISSUE TITLE
User birthday format clarification

### ISSUE THREAD

**Issue by nickma42** [facebook/facebook-php-sdk-v4#570]
The latest graph api docs show that a user's birthday may be in the format MM/DD/YYYY, MM/DD or YYYY depending on what information the user has chosen to share:

https://developers.facebook.com/docs/graph-api/reference/user

> The person's birthday. This is a fixed format string, like MM/DD/YYYY. However, people can control who can see the year they were born separately from the month and day so this string can be only the year (YYYY) or the month + day (MM/DD)

The API code just casts whatever is returned to a DateTime which means I wouldn't be able to tell what bits of information the user had supplied or omitted. If they only supplied the year, the DateTime would default the month and day to the current one. I don't actually know if this applies to apps too or if apps always receive the full birthday.

I couldn't actually figure out how to control what aspects of my birthday could be shared with the app so I couldn't test myself - so any clarification here would be great.

**Comment by SammyK** [comment]
Nice find. This certainly needs to be fixed. Would you like to send in a PR? :)

**Comment by nickma42** [comment]
I'd love to - I'd appreciate any thoughts about the best way to resolve it but I'm leaning towards just returning the field as-is from the API since I can't think of any useful casting that could be done. I think any logic handling the various return formats would always be specific to the use case ... In my particular circumstance we have 3 columns for birth day/month/year.

**Comment by SammyK** [comment]
Yeah, this one is a little tough since we can't change the behavior of what's there without breaking compatibility. Here's how birthday is currently being handled with all the different formats:

<img width="292" alt="screen shot 2016-04-07 at 3 17 06 pm" src="https://cloud.githubusercontent.com/assets/578780/14365546/fdc33628-fcd3-11e5-9812-00dbabbdee72.png">

So any missing pieces will be assumed to be "today".

I propose we add three new methods to `GraphUser`:

```php
$user->getBirthdayMonth();
$user->getBirthdayDay();
$user->getBirthdayYear();
```

Those will return plain `int` or `null`. Then we'll need to update the docs to warn people about the potential wonkiness of `getBirthday()` and describe the new methods as the most reliable method for obtaining the date of birth.

@yguedidi You have any thoughts on this? :)

**Comment by yguedidi** [comment]
I'd suggest to create a class `Birthday`, that simply extends `\DateTime` for BC and will add two methods `hasDate()` and `hasYear()`. Then, use it instead of `\DateTime` when casting :)

**Comment by SammyK** [comment]
Once again @yguedidi has a much better approach than me. :) That's brilliant and we get to keep BC.

**Comment by nickma42** [comment]
Thanks - sounds like an ideal approach. @yguedidi do you have a suggestion as to where I put this Birthday class and what I call it? It doesn't really fit into the existing structure as far as I can tell and I'm not sure it's worth a new folder for 'casted' graph attributes since this would be the only one (except for possibly the AccessToken class but that's already in the Authentication folder) and putting it in /Facebook seems a bit messy.

Other than that I've got a PR ready to go :)

**Comment by yguedidi** [comment]
@nickma42 It's related to nodes, so I think it can go in `GraphNodes` folder for the moment ;)

**Comment by SammyK** [comment]
Closing this since this will be addressed in #573. :)
